checker/src/main/java/org/checkerframework/checker/nullness/CollectionToArrayHeuristics.java

```diff
@@ -132,38 +138,77 @@ private void setComponentNullness(boolean isNonNull, AnnotatedTypeMirror type) {
     }
 
     /**
-     * Returns true if {@code argument} is one of the array creation trees that the heuristic
-     * handles.
+     * Returns {@code true} if {@code argument} is an array argument to {@link
+     * Collection#toArray(Object[]) toArray(T[])} that this heuristic recognizes as producing a
+     * non-null array of the element type.
      *
-     * @param argument the tree passed to {@link Collection#toArray(Object[])
-     *     Collection.toArray(T[])}
-     * @param receiver the expression for the receiver collection
-     * @return true if the argument is handled and assume to return nonnull elements
+     * <p>Recognized forms include:
+     *
+     * <ul>
+     *   <li>{@code new T[] { }} (empty array literal)
+     *   <li>{@code new T[0]} (zero-length allocation)
+     *   <li>{@code new T[receiver.size()]} where {@code receiver} is the {@code Collection} on
+     *       which {@code toArray} is invoked
+     *   <li>an identifier or member select denoting a {@code private final} array field whose
+     *       initializer is an empty array (e.g., {@code private final T[] EMPTY = new T[0];})
+     * </ul>
+     *
+     * @param argument the expression passed as the {@code T[]} parameter to {@code toArray}
+     * @param receiver the simple name of the {@code Collection} receiver used to match {@code new
+     *     T[receiver.size()]}}
+     * @return {@code true} if {@code argument} matches one of the recognized forms
      */
-    private boolean isHandledArrayCreation(Tree argument, String receiver) {
-        if (!(argument instanceof NewArrayTree)) {
+    private boolean isRecognizedToArrayArg(ExpressionTree argument, String receiver) {
+        final NewArrayTree newArr;
+        argument = TreeUtils.withoutParens(argument);
+
+        if (argument instanceof NewArrayTree) {
+            newArr = (NewArrayTree) argument;
+        } else if ((argument instanceof IdentifierTree) || (argument instanceof MemberSelectTree)) {
+            Element el = TreeUtils.elementFromUse(argument);
+            if (el == null || !el.getKind().isField()) {
+                return false;
+            }
+            VariableElement fieldElement = (VariableElement) el;
+            if (!(fieldElement
+                            .getModifiers()
+                            .containsAll(EnumSet.of(Modifier.PRIVATE, Modifier.FINAL))
+                    && ElementUtils.getType(fieldElement).getKind() == TypeKind.ARRAY)) {
+                return false;
+            }
+            Tree declTree = atypeFactory.declarationFromElement(fieldElement);
+            if (declTree == null || !(declTree instanceof VariableTree)) {
```

Comment:
```suggestion
            if (!(declTree instanceof VariableTree)) {
```

Is this equivalent, as instanceof already checks for null?